src/libraries/System.Net.Primitives/src/System/Net/IPNetwork.cs

```diff
@@ -0,0 +1,400 @@
+    /// </remarks>
+    public readonly struct IPNetwork : IEquatable<IPNetwork>, ISpanFormattable, ISpanParsable<IPNetwork>
+    {
+        /// <summary>
+        /// Gets the <see cref="IPAddress"/> that represents the prefix of the network.
+        /// </summary>
+        public IPAddress BaseAddress { get; }
+
+        /// <summary>
+        /// Gets the length of the network prefix in bits.
+        /// </summary>
+        public int PrefixLength { get; }
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="IPNetwork"/> class with the specified <see cref="IPAddress"/> and prefix length.
+        /// </summary>
+        /// <param name="baseAddress">The <see cref="IPAddress"/> that represents the prefix of the network.</param>
+        /// <param name="prefixLength">The length of the prefix in bits.</param>
+        /// <exception cref="ArgumentNullException">The specified <paramref name="baseAddress"/> is <see langword="null"/>.</exception>
+        /// <exception cref="ArgumentOutOfRangeException">The specified <paramref name="prefixLength"/> is smaller than `0` or longer than maximum length of <paramref name="prefixLength"/>'s <see cref="AddressFamily"/>.</exception>
+        /// <exception cref="ArgumentException">The specified <paramref name="baseAddress"/> has non-zero bits after the network prefix.</exception>
+        public IPNetwork(IPAddress baseAddress, int prefixLength)
+        {
+            ArgumentNullException.ThrowIfNull(baseAddress);
+            Validate(baseAddress, prefixLength, throwOnFailure: true);
+
+            BaseAddress = baseAddress;
+            PrefixLength = prefixLength;
+        }
+
+        // Non-validating ctor
+        private IPNetwork(IPAddress baseAddress, int prefixLength, bool _)
+        {
+            BaseAddress = baseAddress;
+            PrefixLength = prefixLength;
+        }
+
+        /// <summary>
+        /// Determines whether a given <see cref="IPAddress"/> is part of the network.
+        /// </summary>
+        /// <param name="address">The <see cref="IPAddress"/> to check.</param>
+        /// <returns><see langword="true"/> if the <see cref="IPAddress"/> is part of the network; otherwise, <see langword="false"/>.</returns>
+        /// <exception cref="ArgumentNullException">The specified <paramref name="address"/> is <see langword="null"/>.</exception>
+        public bool Contains(IPAddress address)
+        {
+            ArgumentNullException.ThrowIfNull(address);
+
+            if (BaseAddress == null)
+            {
+                throw new InvalidOperationException(SR.net_uninitialized_ip_network_instance);
+            }
+
+            if (address.AddressFamily != BaseAddress.AddressFamily)
+            {
+                return false;
+            }
+
+            if (PrefixLength == 0)
+            {
+                return true;
+            }
```

Comment:
Is this common enough to be worth special-casing?  Is there real-world use that has a prefix length of 0?

Reply:
The latter. It prevents the `uint.MaxValue << 32` and `UInt128.MaxValue << 128` special cases in the code below.